Orange/widgets/widget.py

```diff
@@ -539,6 +539,48 @@ def progressBarFinished(self, processEvents=QEventLoop.AllEvents):
         if processEvents is not None and processEvents is not False:
             qApp.processEvents(processEvents)
 
+    def progressBar(self, show_bar=True):
+        """
+        Decorator and context manager for progress bar.
+
+        Using this decorator ensures that the progress bar is removed at the
+        end of the with-block or at exit from function even in case of
+        exceptions, therefore removing the need to remove the bar in
+        `finally` blocks.
+
+        Usage as context handler:
+
+            with self.progressBar:
+                do_something()
+                update_progress_bar_while_doing_it()
+
+        Usage as decorator:
+
+            @OWWidget.progressBar
+            def compute(self):
+                do_something()
+                update_progress_bar_while_doing_it()
```

Comment:
I don't like this usage. It removes the progress bar initialization logic away from the code and into a decorator. This implies that either the code with a call to `progressBarSet()` (the example in this docstring could be clearer in this regard — what's `update_progress_bar_while_doing_it()`?), or the decorator could both be used on their own, which is obviously not the case.

There should be one obvious way to do it. Nothing wrong with `with self.progressBar()` (other than it increments indent level, but so does `finally`).